File: swift/242-Valid-Anagram.swift
/**
 * Question Link: https://leetcode.com/problems/valid-anagram/
 */

class ValidAnagram {
    func isAnagram(_ s: String, _ t: String) -> Bool {
      if s.count != t.count {
        return false
      }
      var countS = [Int:Int]()
      var countT = [Int:Int]()
      let sChars = Array(s)
      let tChars = Array(t)
      for i in 0..<sChars.count {
          countS[Int(sChars[i].asciiValue!)] = 1 + countS[Int(sChars[i].asciiValue!), default: 0]
          countT[Int(tChars[i].asciiValue!)] = 1 + countT[Int(tChars[i].asciiValue!), default: 0]
      }
      for (key, _) in countS {
          if countS[key] != countT[key, default: 0] {
              return false
          }
      }
      return true
    }
    
    func isAnagramByLengthAndSortCheck(_ s: String, _ t: String) -> Bool {
        
        // if the String lengths don't match, then they cannot be anagrams of each other
        guard s.count == t.count else { return false }
        
        // sorting the String Characters puts them in the same order, which allows us to check equality
        guard s.sorted() == t.sorted() else { return false }
        
        // if both guards succeed, then we have an anagram
        return true
    }
}

File: go/01_arrays_and_hashing/242_Valid_Anagram/242_Valid_Anagram.go
package isAnagram

func isAnagram(s string, t string) bool {
	characters := make(map[byte]int)

	if len(s) != len(t) {
		return false
	}

	for i := 0; i < len(s); i++ {
		_, ok := characters[s[i]]
		if !ok {
			characters[s[i]] = 1
		} else {
			characters[s[i]] += 1
		}

	}

	for i := 0; i < len(t); i++ {
		val, ok := characters[t[i]]
		if ok && val > 0 {
			characters[t[i]] -= 1
		} else {
			return false
		}
	}

	return true

}
